allensdk/brain_observatory/ecephys/current_source_density/_current_source_density.py

```diff
@@ -189,102 +135,57 @@ def accumulate_lfp_data(timestamps, lfp_raw, lfp_channels, trial_windows, extrac
-    '''
-    return RegularGridInterpolator((timestamps,), lfp_raw[:, channel], method=method)
+    msg = 'extracted lfp data for {} trials, {} channels, and {} samples'
+    logging.info(msg.format(*accumulated.shape))
+    return accumulated * volts_per_bit
+
+
+def compute_csd(lfp_stack: np.ndarray,
+                spacing: float) -> Tuple[np.ndarray, np.ndarray]:
+    '''Compute current source density for real or virtual channels from
+    a neuropixels probe.
 
-    
-def compute_csd(accumulated_lfp, channels=None, missing_channels=None, spacing=0.04):
-    ''' Compute current source density for a subset of channels on a neuropixels probe.
+    Uses the numerical approximation to the Laplacian, after Pitts (1952).
 
     Parameters
     ----------
-    accumulated_lfp : numpy.ndarray
-        Stack of extracted LFP traces surrounding presentation of a common stimulus. Dimensions are trials X channels X samples.
-        
-    spacing: float
-        vertical spacing between channels, in milimeter.
-        TODO: spacing needs to be inferred through LFP channel id spacing
+    lfp_stack: numpy.ndarray
+        Stack of LFP traces surrounding presentation of a common stimulus.
+        Dimensions are trials X channels X samples.
+    spacing : float
+        Distance between channels, in millimeters. This spacing may be
+        physical distances between channels or a virtual distance if channels
+        have been interpolated to new virtual positions.
 
     Returns
     -------
-    csd : numpy.ndarray
-        current source density. Dimensions are channels X samples
-
+    Tuple[csd, csd_channels]:
+        csd : numpy.ndarray
+            Current source density. Dimensions are channels X samples.
+        csd_channels: numpy.ndarray
+            Array of channel indices for CSD.
     '''
 
-    accumulated_lfp = accumulated_lfp - np.mean(accumulated_lfp, axis=2)[:, :, None]
-    lfp = np.mean(accumulated_lfp, axis=0)
+    # mean_lfp = np.zeros((np.shape(lfp_stack)[1] + 2,
+    #                      np.shape(lfp_stack)[2]))
 
-    lfp, channels = resample_to_regular(lfp, channels, missing_channels)
+    # mean_lfp[1:-1,:] = np.nanmean(lfp_stack, axis=0)
+    # mean_lfp[0,:] = mean_lfp[1,:]
+    # mean_lfp[-1,:] = mean_lfp[-2,:]
 
-    lfp_smooth=np.zeros([np.shape(lfp)[0] - 2, np.shape(lfp)[1]])
-    csd = np.zeros([np.shape(lfp_smooth)[0] - 2, np.shape(lfp_smooth)[1]])
-    
-    for t in range(np.shape(lfp)[1]):  # time
-        for d in range(1, np.shape(lfp)[0] - 1):  # depth
-            lfp_smooth[d - 1, t] = 0.25 * ( lfp[d - 1, t] + 2 * lfp[d, t] + lfp[d + 1, t] )
-    
-    for t in range(np.shape(lfp_smooth)[1]): # time
-        for d in np.arange(1, np.shape(lfp_smooth)[0] - 1):
-            csd[d - 1, t]=(1 / spacing ** 2) * ( lfp_smooth[d + 1, t] - 2 * lfp_smooth[d, t] + lfp_smooth[d - 1, t] )    
-    
-    return csd, channels
+    mean_lfp = np.nanmean(lfp_stack, axis=0)
```

Comment:
You average out LFP over trials here. I would pull this out of the csd calculation, since you are doing processing of LFP. Is this averaging over trials corresponding to the same stimulus?

Reply:
Moved trial averaging outside of `compute_csd()`.